### PR TITLE
Pbinder/auto resume

### PR DESCRIPTION
**Pull request by polinabinder1**
PEFT checkpointing and inference for esm2.

### CODE REVIEW

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 1]
```diff
@@ -386,66 +404,71 @@ def finetune_esm2_entrypoint():
...
-        overlap_param_gather=not args.no_overlap_param_gather,
-        average_in_collective=not args.no_average_in_collective,
-        grad_reduce_in_fp32=args.grad_reduce_in_fp32,
-        ckpt_async_save=not args.avoid_ckpt_async_save,
-        label_column=args.label_column,
-    )
+    if args.lora_checkpoint_path and not args.lora_finetune:
+        parser.error("Arguments --lora=checkpoint-path cannot be set when not using lora-finetune.")
+
+    with megatron_parallel_state_utils.distributed_model_parallel_state(43):
```
Comment (trvachov) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:410]: should this seed be surfaced somewhere?

**Comment by trvachov** [comment]
The resulting model finetuned with LoRA has this file...do you know what it's supposed to have in it?
```
cat /results/lora_model/checkpoint-step\=999-consumed_samples\=64000.0-last/weights/adapter_metadata.json
{"model_ckpt_path": "None"}
```

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 2]
```diff
@@ -334,6 +353,7 @@ def train_model(
             weight_decay=0.01,
             adam_beta1=0.9,
             adam_beta2=0.98,
+            clip_grad=1.0,  # Add gradient clipping
```
Comment (trvachov) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:356]: Interesting -- why do we need to add gradient clipping to enable LoRA?

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 3]
```diff
@@ -618,7 +650,7 @@ def get_parser():
         "--val-check-interval",
         type=int,
         required=False,
-        default=10000,
+        default=5,
```
Comment (trvachov) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:653]: Are these tiny defaults reasonable?

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/conftest.py [hunk 4]
```diff
@@ -109,3 +127,42 @@ def dummy_protein_sequences(dummy_data_per_token_classification_ft):
...
+    Returns:
+        Callable: A function that takes a checkpoint directory path and returns the loaded state dict.
+    """
+    if not torch.cuda.is_available():
+        pytest.skip("Distributed checkpoint loading requires CUDA")
+
+    def _load_dcp(ckpt_dir):
+        from pathlib import Path
+
+        import torch
```
Comment (trvachov) [inline sub-packages/bionemo-esm2/tests/bionemo/esm2/conftest.py:145]: duplicate import?

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_infer_esm2.py [hunk 5]
```diff
@@ -53,23 +54,32 @@ def padded_tokenized_sequences(dummy_protein_sequences):
...
     padded_tokenized_sequences,
+    config_class,
 ):
+    checkpoint_path = load("esm2/8m:2.0")
     data_path = dummy_protein_csv
     result_dir = tmpdir / "results"
     min_seq_len = 1024  # Minimum length of the output batch; tensors will be padded to this length.
-
+    if with_peft:
+        lora_checkpoint_path = load("esm2/esm2_finetuned_peft_weights:1.0", source="pbss") / "weights"
```
Comment (trvachov) [inline sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_infer_esm2.py:77]: I think we would like all tests to be runnable by the public, Ie. this must be on NGC

**Review by trvachov — CHANGES_REQUESTED** [review]

**Comment by trvachov** [comment]
Aside from small comments here and there, my main concern is we need to put the test data on NGC, because as is this command fails `download_bionemo_data esm2/esm2_finetuned_peft_weights:1.0`

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py (hunk 1 above)
Comment (polinabinder1) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py]: This is fixed

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/peft.py [hunk 6]
```diff
@@ -37,13 +40,66 @@
...
+        self.wrapped_io.adapter_ckpt_path = self.peft_ckpt_path
+
+    def on_predict_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
+        """Event hook.
+
+        Args:
+            trainer: The trainer object.
+            pl_module: The LightningModule object.
+        """
+        self._maybe_apply_transform(trainer)
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/peft.py:79]: In the model transform class: https://github.com/NVIDIA/NeMo/blob/08c4da837b9ff476d662669353d60d8a71550eae/nemo/lightning/pytorch/callbacks/model_transform.py#L85

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 7]
```diff
@@ -703,6 +735,22 @@ def get_parser():
...
+        default=False,
+        help="Perform fine-tuning with LoRA.",
+    )
+
+    parser.add_argument(
+        "--lora-checkpoint-path",
+        type=str,
+        required=False,
+        default=None,
+        help="Path to the lora states to restore from.",
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:751]: done

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py (hunk 2 above)
Comment (polinabinder1) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:356]: removing it

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py (hunk 3 above)
Comment (polinabinder1) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:653]: no they're not.

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_infer_esm2.py (hunk 5 above)
Comment (polinabinder1) [reply sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_infer_esm2.py:77]: working on that!

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 8]
```diff
@@ -35,13 +36,15 @@
...
 )
+from bionemo.esm2.model.finetune.peft import ESM2LoRA
 from bionemo.esm2.model.finetune.sequence_model import ESM2FineTuneSeqConfig
 from bionemo.esm2.model.finetune.token_model import ESM2FineTuneTokenConfig
 from bionemo.llm.model.biobert.lightning import biobert_lightning_module
 from bionemo.llm.model.biobert.model import BioBertConfig
 from bionemo.llm.model.config import TorchmetricsConfig
 from bionemo.llm.utils.datamodule_utils import float_or_int_or_none, infer_global_batch_size
 from bionemo.llm.utils.logger_utils import WandbConfig, setup_nemo_lightning_logger
+from bionemo.testing import megatron_parallel_state_utils
```
Comment (farhadrgh) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:47]: I am pretty sure this script cannot depend on testing. Why would we need that?

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/peft.py [hunk 9]
```diff
@@ -37,13 +40,66 @@
...
+        peft_ckpt_path: Optional[str] = None,
+        freeze_modules: List[str] = ["encoder", "embedding"],
+        *args,
+        **kwarg,
+    ):
+        """Initialize the LoRA Adapter.
+
+        Args:
+            peft_ckpt_path: config for peft chekpoint.
+            freeze_modules: modules to freeze.
```
Comment (farhadrgh) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/peft.py:54]: Is this overriding the encoder_frozen argument in finetune configs?

**Review by farhadrgh — APPROVED** [review]
LGTM

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/conftest.py (hunk 4 above)
Comment (trvachov) [reply sub-packages/bionemo-esm2/tests/bionemo/esm2/conftest.py]: Resolved, I guess?

**Comment by trvachov** [comment]
Approving under the assumption the NGC data is upcoming soon :)

**Review by trvachov — APPROVED** [review]
